pytest_mh/utils/firewall.py

```diff
@@ -88,8 +82,145 @@ def parse_port_spec(self, spec: PortSpec) -> tuple[int, ProtocolSpec]:
 
         return (port, protocol)
 
+    def parse_host_spec(self, host: HostSpec) -> str:
+        """
+        Parse host specification into hostname.
+
+        :raises ValueError: When invalid type is given.
+        :return: Hostname.
+        :rtype: str
+        """
+        if isinstance(host, str):
+            return host
+        elif isinstance(host, MultihostHost):
+            return host.hostname
+        elif isinstance(host, MultihostRole):
+            return host.host.hostname
+        else:
+            raise ValueError(f"Unexpected type of host: {type(host)}")
+
+
+class FirewallInboundRules(object):
+    """
+    Configure firewall inbound rules.
+
+    All changes are automatically reverted when a test is finished.
+    """
+
+    @abstractmethod
+    def accept_port(self, port: PortSpec | list[PortSpec]) -> None:
+        """
+        Accept incoming traffic on given port.
+
+        .. code-block:: python
+
+            firewall.inbound.accept_port(389)  # missing protocol defaults to "tcp"
+            firewall.inbound.accept_port((389, "tcp"))
+            firewall.inbound.accept_port([389, 636])
+
+        :param port: Port (integer), (port, protocol) tuple.
+        :type port: PortSpec | list[PortSpec]
+        """
+        pass
+
+    @abstractmethod
+    def drop_port(self, port: PortSpec | list[PortSpec]) -> None:
+        """
+        Drop incoming traffic on given port.
+
+        .. code-block:: python
+
+            firewall.inbound.drop_port(389)  # missing protocol defaults to "tcp"
+            firewall.inbound.drop_port((389, "tcp"))
+            firewall.inbound.drop_port([389, 636])
+
+        :param port: Port (integer), (port, protocol) tuple.
+        :type port: PortSpec | list[PortSpec]
+        """
+        pass
+
```

Comment:
No `reject_port`?

Reply:
No, Windows Firewall does not support reject, only drop.

Reply:
I can't think of a QE scenario in which we distinguish the difference between dropping and rejecting traffic, just as long as it's dropped, which both do. I think one function is better.

Reply:
reject = return immediately, useful for testing sssd in offline mode and you don't have to wait for timeout
drop = you must wait for timeout to bring sssd offline, useful when testing timeouts

Reply:
We should have reject here, in the API as an abstract method. Windows fw will ether not implement it or just use drop in implementation. Rejecting traffic is useful to have test faster (reject = return immediately).

Reply:
It is not needed to have reject in abstract class. You will have it if needed on linux providers, just not on AD. It is also inbound traffic, that you are not going to use anyway. Outbound rules will be called from client where reject is available.

The exception is raised automatically if method is not provided via @abstractmethod.

Reply:
That's good to know, and it makes sense. Because rejecting the server responds with access denied,  and dropping won't respond at all. So the only one we'd need is drop.

Reply:
To clarify further, reject is not firewall thing. You can do it in iptables and we can use it to avoid waiting for timeouts during testing, but it is actually for routers. Implementing Window's reject and drop would mess up expectations about the test and since we can't implement it everywhere, it should not be part of generic API.

But really, what we want to use mostly are the outbound rules (as was requested on the mailing list and also in some pull request) on the linux client which will provide all three: accept, drop and reject. So calling `client.firewall.outbound.reject_host(provider)` will work for all providers as expected.